### PR TITLE
Added mverteuil keymaps and userspace common code

### PR DESCRIPTION
**Pull request by mverteuil**
## Description

Adds keymap/userspace

- mverteuil userspace
- Knops mini keymap
- Preonic keymaps
  - 1x2u
  - 2x2u

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

N/A

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/knops/mini/keymaps/mverteuil/config.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef CONFIG_USER_H
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/config.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

File: keyboards/knops/mini/keymaps/mverteuil/config.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#define CONFIG_USER_H
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/config.h]:
```suggestion
```

File: keyboards/knops/mini/keymaps/mverteuil/config.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+#if defined(AUDIO_ENABLE)
+    #define AUDIO_CLICKY
+    #define B5_AUDIO
+    #define C6_AUDIO
+    #define STARTUP_SONG SONG(ZELDA_PUZZLE)
+    #define GOODBYE_SONG SONG(COIN_SOUND)
+#endif
+#endif
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/config.h]:
```suggestion
```

File: keyboards/preonic/keymaps/mverteuil/keymap.c [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+            if (clockwise) {
+                muse_tempo += 1;
+            } else {
+                muse_tempo -= 1;
+            }
+        }
+    } else {
+        if (clockwise) {
+            register_code(KC_PGDN);
+            unregister_code(KC_PGDN);
```
Comment (drashna) [inline keyboards/preonic/keymaps/mverteuil/keymap.c]: This can be replaced with `tap_code(KC_PGDN)`, as it will do both of these (plus an optional delay between the two)

File: users/mverteuil/mverteuil.c [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#include QMK_KEYBOARD_H
+#include "mverteuil.h"
+
+void register_left_brace() {
+    register_code(KC_LSHIFT);
+    register_code(KC_LBRACKET);
```
Comment (drashna) [inline users/mverteuil/mverteuil.c]: This can be replaced by: 
```c
register_code16(LSFT(KC_LBRACKET));
```
or
```c
register_code16(KC_LEFT_CURLY_BRACE);
```

https://github.com/qmk/qmk_firmware/blob/master/quantum/quantum_keycodes.h#L550-L612

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 6]
```diff
@@ -0,0 +1,349 @@
+#include QMK_KEYBOARD_H
+
+#define MAXIMIZ A(G(KC_F))
+#define NXTDSPL C(A(G(KC_RIGHT)))
+#define BTMHALF A(G(KC_DOWN))
+#define TOPHALF A(G(KC_UP))
+#define NXTTHRD SS_DOWN(X_LCTRL) SS_DOWN(X_LALT) SS_TAP(X_RIGHT) SS_UP(X_LALT) SS_UP(X_LCTRL)
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
#define NXTTHRD SS_LCTL(SS_LALT(SS_TAP(X_RIGHT)))
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+ *	 < 0 >   < 1 >   < 2 >     <---      These front-LEDs are called 'Layer LEDs'
+ *							             To turn one of them on, use:
+ *										 set_layer_led( [0-2] );
+ */
+
+void set_switch_led(int ledId, bool state) {
+    if (state) {
+        switch (ledId) {
+            case 1:
+                PORTD |= (1 << 7);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]: These can be simplified to eg.
```suggestion
                writePinHigh(D7);
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+                break;
+            case 2:
+                if ((PINB & (1 << 7)) != 0) {
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
                if (readPin(B7)) {
```

File: users/mverteuil/mverteuil.c (hunk 5 above)
Comment (mverteuil) [reply users/mverteuil/mverteuil.c]: AHHHHH man I love you! I can just use that directly in `keymap.c` and skip the wrappers altogether ;)

File: keyboards/preonic/config.h [hunk 9]
```diff
@@ -41,6 +41,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #define QMK_LED     E6
 #define QMK_SPEAKER C6
 
+#define AUDIO_CLICKY
```
Comment (fauxpark) [inline keyboards/preonic/config.h]: Please put this in your own config.h instead.

File: keyboards/knops/mini/mini.h [hunk 10]
```diff
@@ -23,8 +23,9 @@
 // The first section contains all of the arguments
 // The second converts the arguments into a two-dimensional array
 #define LAYOUT( \
-	K00, K01, K02, K03, K04, K05  \
+	K00, K01, K02, \
```
Comment (fauxpark) [inline keyboards/knops/mini/mini.h]: Please undo the changes to this file.

**Comment by mverteuil** [comment]
Thanks for all of the great feedback, I will integrate your changes and push shortly! ❤️ 👍

File: users/mverteuil/mverteuil.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+#endif
+
+/* Keypad aliases */
+#define KC_DOLR KC_DOLLAR
+#define KC_NAST KC_KP_ASTERISK
+#define KC_NMIN KC_KP_MINUS
+#define KC_NPLU KC_KP_PLUS
+#define KC_NENT KC_KP_ENTER
+#define KC_NDOT KC_KP_DOT
+#define KC_PCNT KC_PERCENT
```
Comment (fauxpark) [inline users/mverteuil/mverteuil.h]: These already have 7-character aliases defined in core.

File: keyboards/preonic/keymaps/mverteuil/keymap.c [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |NP FRw| Ctrl | Alt  | GUI  |Lower |    Space    |Raise | /  ? | Left | Down |Right |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_QWERTY] = LAYOUT_preonic_1x2uC ( \
+  KC_GRV,  KC_EXLM, KC_AT,   KC_HASH, KC_DOLR, KC_PCNT, KC_CIRC, KC_AMPR, KC_ASTR, TD_USMI, TD_PLEQ, KC_BSPC, \
+  KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    TD_SQDQ, \
+  TD_CLES, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    TD_SCOL, KC_ENT,  \
+  KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_UP,   KC_RGUI, \
+  TD_NPFR, KC_LCTL, KC_LALT, KC_LGUI, LOWER,        KC_SPC,      RAISE,   TD_SLQU, KC_LEFT, KC_DOWN, KC_RGHT  \
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mverteuil/keymap.c]: The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

File: keyboards/preonic/keymaps/mverteuil/keymap.c [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+            layer_on(_NUMPAD);
+            PLAY_SONG(s_numpad_toggle);
+            break;
+        case DOUBLE_HOLD:
+            layer_on(_FUNCTIONROW);
+            break;
+        case DOUBLE_TAP:
+            if (layer_state_is(_FUNCTIONROW)) {
+                layer_off(_FUNCTIONROW);
+                ;
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mverteuil/keymap.c]:
```suggestion
```

**Comment by mverteuil** [comment]
@fauxpark I've implemented your requests, thanks for the tips!

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+}
+
+// Runs state check and changes underglow color and animation
+layer_state_t default_layer_state_set_user(layer_state_t state) { return state; }
+
+void matrix_init_kb(void) {
+    led_init_ports();
+
+    writePinHigh(B7);
+    DDRB &= ~(1 << 7);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
    setPinInputHigh(B7);
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+    return state;
+}
+
+// Runs state check and changes underglow color and animation
+layer_state_t default_layer_state_set_user(layer_state_t state) { return state; }
+
+void matrix_init_kb(void) {
+    led_init_ports();
+
+    writePinHigh(B7);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+            set_switch_led(5, true);
+            set_switch_led(6, true);
+            break;
+    }
+}
+
+void led_init_ports() {
+    writePinHigh(B7);
+
+    DDRB &= ~(1 << 7);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
    setPinInputHigh(B7);
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+            set_switch_led(3, true);
+            set_switch_led(4, true);
+            set_switch_led(5, true);
+            set_switch_led(6, true);
+            break;
+    }
+}
+
+void led_init_ports() {
+    writePinHigh(B7);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+    DDRB &= ~(1 << 7);
+
+    DDRD |= (1 << 7);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
    setPinOutput(D7);
```
etc.

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+    DDRE |= (1 << 6);
+    DDRB |= (1 << 4);
+    DDRD |= (1 << 6);
+
+    DDRD |= (1 << 5);
+    DDRB |= (1 << 6);
+    DDRB |= (1 << 0);;
+
+    led_set_layer(0);
+    _delay_ms(200);
```
Comment (fauxpark) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
    wait_ms(200);
```
etc.

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 16 above)
Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Thanks! I didn't know what the alias for setting/reading DDRB pins were, it's been a long time since I did any microcontroller stuff and much of the source related to LEDs comes from the default keymap. 

I feel I'd made some minor improvements, but these are way more readable.

Should we also add these improvements to the default keymap, or there code ownership concerns with that?

Now, if I only understood why the initialization code is the way it is... there's no docs from Knops w.r.t. LEDs, so I figured most of it out by trial and error and experimenting/discovery with removing certain calls.

I read something about this device essentially being an Arduino Leonardo, so I wonder if that's where I should be looking...

In any event, I'll integrate these suggestions and thanks again. 👍

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 20 above)
Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: I should probably just replace this with a loop

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 17 above)
Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Ahh, I see this is actually missing from the two defaults now, I think I had taken it from https://github.com/qmk/qmk_firmware/blob/master/keyboards/knops/mini/keymaps/knops/keymap.c#L86

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 16 above)
Comment (fauxpark) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Ah, I didn't realise this was from the default keymap - that's fine to do the same there, but I think I'd prefer a separate PR for it, as this one's quite large already.

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+                writePinLow(E6);
+                break;
+            case 5:
+                writePinLow(B4);
+                break;
+            case 6:
+                writePinLow(D6);
+                break;
+        }
+    }
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]: For this entire thing, you could use `writePin(xx, state)` instead, I think.  https://docs.qmk.fm/#/internals_gpio_control

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+    _delay_ms(200);
+    led_set_layer(2);
+    _delay_ms(200);
+    led_set_layer(0);
+}
+
+// on layer change, no matter where the change was initiated
+// Then runs keymap's layer change check
+layer_state_t layer_state_set_user(layer_state_t state) {
+    // There is no layer "0", it's always at least 1
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Not entirely true. Layer 0 is the default layer.

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 23]
```diff
@@ -0,0 +1,349 @@
...
+    DDRD |= (1 << 6);
+
+    DDRD |= (1 << 5);
+    DDRB |= (1 << 6);
+    DDRB |= (1 << 0);;
+
+    led_set_layer(0);
+    wait_ms(200);
+    led_set_layer(1);
+    _delay_ms(200);
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
    wait_ms(200);
```

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c [hunk 24]
```diff
@@ -0,0 +1,349 @@
...
+    led_set_layer(2);
+    _delay_ms(200);
```
Comment (drashna) [inline keyboards/knops/mini/keymaps/mverteuil/keymap.c]:
```suggestion
    wait_ms(200);
```

**Comment by fauxpark** [comment]
By the way, you don't need to keep force-pushing - we squash all PRs anyway.

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 22 above)
Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Yeah I noticed that and revised the comment, not sure why I never see layer 1 off though

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 21 above)
Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: yes, except for the B7/C6/C7 oddness— not sure why it changes based on the state of B7 🤔

**Comment by mverteuil** [comment]
> By the way, you don't need to keep force-pushing - we squash all PRs anyway.

Fair enough, just a force of habit because it's easier to rebase off master on other projects when I remove fixup commits ;)

File: keyboards/knops/mini/keymaps/mverteuil/keymap.c (hunk 21 above)
Comment (drashna) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Well, you could still do so. 

And that block could be: 
```c
writePinLow( readPin(B7) ? C6 : C7, state);
```

I'm not sure that would work, but ... "looks right".

Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Hmm, the device is on my desk at my office— so I will validate tomorrow

Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Generally if I encounter a block like this in python, I will separate the concerns by using a dict mapping switch to pin; is there a similar idiom in C where I can separate the data mapping from the data processing concern?

i.e. can I use something more compact or expressive than a switch/case to get the correct pin from the switch id?

Comment (drashna) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Honestly, I'm not sure.

Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: Fair enough, just a C noob looking to take any help I can get ;)

Comment (mverteuil) [reply keyboards/knops/mini/keymaps/mverteuil/keymap.c]: This works 👍

**Comment by mverteuil** [comment]
Ok I think this should be good for review again

File: .gitignore [hunk 25]
```diff
@@ -55,6 +55,7 @@ util/Win_Check_Output.txt
 .vscode/last.sql
 .vscode/temp.sql
 .vscode/ipch/
+.vscfavoriterc
```
Comment (drashna) [inline .gitignore]: Could this be removed/reverted? 

If it's something that should be added, it should be done so in a separate PR from the keymaps

Comment (mverteuil) [reply .gitignore]: sure, I can just move it to my `~/.config/git/ignore` for now

**Comment by mverteuil** [comment]
Thanks again for all the help and feedback y'all 💟

**Comment by drashna** [comment]
Thanks for contributing!
